### PR TITLE
Regr/period range large value/issue 36430

### LINKED ISSUE #32749: BUG: regression in error raised by idxmin/idxmax for extension dtypes
With pandas 1.0.1:

```
In [7]: from pandas.tests.extension.decimal import DecimalArray, make_data 

In [8]: s = pd.Series(DecimalArray(make_data()[:5]))  

In [9]: s 
Out[9]: 
0    Decimal: 0.25866656130631138221787068687262944...
1    Decimal: 0.11511905452780368808163302674074657...
2    Decimal: 0.15301679241167220890673661415348760...
3    Decimal: 0.41125672759464626526693109553889371...
4    Decimal: 0.46391316685725048074573351186700165...
dtype: decimal

In [10]: s.idxmin()   
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-10-5060aab05d12> in <module>
----> 1 s.idxmin()

~/miniconda3/envs/pandas10/lib/python3.8/site-packages/pandas/core/series.py in idxmin(self, axis, skipna, *args, **kwargs)
   2037         """
   2038         skipna = nv.validate_argmin_with_skipna(skipna, args, kwargs)
-> 2039         i = nanops.nanargmin(com.values_from_object(self), skipna=skipna)
   2040         if i == -1:
   2041             return np.nan

~/miniconda3/envs/pandas10/lib/python3.8/site-packages/pandas/core/nanops.py in _f(*args, **kwargs)
     62             if any(self.check(obj) for obj in obj_iter):
     63                 f_name = f.__name__.replace("nan", "")
---> 64                 raise TypeError(
     65                     f"reduction operation '{f_name}' not allowed for this dtype"
     66                 )

TypeError: reduction operation 'argmin' not allowed for this dtype
```

Now with master you get:

```
In [4]: s.idxmin() 
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-4-7e62303a9d43> in <module>
----> 1 s.idxmin()

~/scipy/pandas/pandas/core/series.py in idxmin(self, axis, skipna, *args, **kwargs)
   1988         """
   1989         skipna = nv.validate_argmin_with_skipna(skipna, args, kwargs)
-> 1990         i = nanops.nanargmin(self._values, skipna=skipna)
   1991         if i == -1:
   1992             return np.nan

~/scipy/pandas/pandas/core/nanops.py in _f(*args, **kwargs)
     69             try:
     70                 with np.errstate(invalid="ignore"):
---> 71                     return f(*args, **kwargs)
     72             except ValueError as e:
     73                 # we want to transform an object array

~/scipy/pandas/pandas/core/nanops.py in nanargmin(values, axis, skipna, mask)
    943     """
    944     values, mask, dtype, _, _ = _get_values(
--> 945         values, True, fill_value_typ="+inf", mask=mask
    946     )
    947     result = values.argmin(axis)

~/scipy/pandas/pandas/core/nanops.py in _get_values(values, skipna, fill_value, fill_value_typ, mask)
    311         # promote if needed
    312         else:
--> 313             values, _ = maybe_upcast_putmask(values, mask, fill_value)
    314 
    315     # return a platform independent precision dtype

~/scipy/pandas/pandas/core/dtypes/cast.py in maybe_upcast_putmask(result, mask, other)
    278     """
    279     if not isinstance(result, np.ndarray):
--> 280         raise ValueError("The result input must be a ndarray.")
    281     if not is_scalar(other):
    282         # We _could_ support non-scalar other, but until we have a compelling

ValueError: The result input must be a ndarray.
```

@jbrockmendel this is from no longer using `values_from_object`, but passing `_values` to the nanops function, while this expects to always receive an ndarray

### LINKED ISSUE #36430: REGR: period_range giving incorrect values for large datetimes
From the pandas-dev mailing list:

> I encountered a bug in pandas 1.1.0 after updating from 1.0.5 that I wanted to pass along.  If a period range is created that extends past the maximum timestamp date of 2262-04-11, then the period range is shifted back by an hour after this date.  The screenshots below show an example with the shifted times highlighted in yellow.

```python
import pandas as pd
from datetime import datetime

start = datetime(2000, 1, 1)
end = datetime(2500, 1, 1)

idx = pd.period_range(start, end, freq="6H")
assert idx[-1].hour % 2 == 0  # raises

idx
# PeriodIndex(['2000-01-01 00:00', '2000-01-01 06:00', '2000-01-01 12:00',
#              '2000-01-01 18:00', '2000-01-02 00:00', '2000-01-02 06:00',
#              '2000-01-02 12:00', '2000-01-02 18:00', '2000-01-03 00:00',
#              '2000-01-03 06:00',
#              ...
#              '2499-12-29 17:00', '2499-12-30 23:00', '2499-12-30 05:00',
#              '2499-12-30 11:00', '2499-12-30 17:00', '2499-12-31 23:00',
#              '2499-12-31 05:00', '2499-12-31 11:00', '2499-12-31 17:00',
#              '2500-01-01 23:00'],
#             dtype='period[6H]', length=730489, freq='6H')

pd.__version__
# '1.1.0'
```

### PR DESCRIPTION
**Pull request by nrebena**
#  Checklist

- [x] closes #36430 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

# Solution

Culprit was the multiplication `unix_date * 24 * 3600 * 10**9 / factor`, for 
```
unix_date = 106752
np.log2(unix_date * 24 * 3600 * 10**9)
# 63.00000011936912
```
That probably lead to an integer overflow somewhere and the observed behaviours.

Splitting the multiplication did the trick.

### CODE REVIEW

File: doc/source/whatsnew/v1.1.3.rst [hunk 1]
```diff
@@ -33,9 +33,10 @@ Fixed regressions
 - Fixed regression in :class:`IntegerArray` unary plus and minus operations raising a ``TypeError`` (:issue:`36063`)
 - Fixed regression in :meth:`Series.__getitem__` incorrectly raising when the input was a tuple (:issue:`35534`)
 - Fixed regression in :meth:`Series.__getitem__` incorrectly raising when the input was a frozenset (:issue:`35747`)
-- Fixed regression in :meth:`read_excel` with ``engine="odf"`` caused ``UnboundLocalError`` in some cases where cells had nested child nodes (:issue:`36122`,:issue:`35802`)
-- Fixed regression in :class:`DataFrame` and :class:`Series` comparisons between numeric arrays and strings (:issue:`35700`,:issue:`36377`)
+- Fixed regression in :meth:`read_excel` with ``engine="odf"`` caused ``UnboundLocalError`` in some cases where cells had nested child nodes (:issue:`36122`, :issue:`35802`)
```
Comment (WillAyd) [inline doc/source/whatsnew/v1.1.3.rst:36]: Can you revert this? Minor nit but it's confusing to include here; can just do a separate PR to clean whatsnew

File: pandas/_libs/tslibs/period.pyx [hunk 2]
```diff
@@ -886,7 +886,10 @@ cdef int64_t get_time_nanos(int freq, int64_t unix_date, int64_t ordinal) nogil:
         # We must have freq == FR_HR
         factor = 10**9 * 3600
 
-    sub = ordinal - unix_date * 24 * 3600 * 10**9 / factor
+    # Fix issue #36430
+    nanos_in_day = 24 * 3600 * 10**9
```
Comment (WillAyd) [inline pandas/_libs/tslibs/period.pyx:890]: Can you add the appropriate suffixes to the constants here? This seems suspect that it would make a difference at all; wonder if the suffixes alone would fix

**Review by WillAyd — CHANGES_REQUESTED** [review]
Thanks for root causing - pretty tricky

**Comment by jreback** [comment]
looks good ping when addressed @WillAyd comments and green.

File: pandas/_libs/tslibs/period.pyx [hunk 3]
```diff
@@ -886,7 +886,10 @@ cdef int64_t get_time_nanos(int freq, int64_t unix_date, int64_t ordinal) nogil:
         # We must have freq == FR_HR
         factor = 10**9 * 3600
 
-    sub = ordinal - unix_date * 24 * 3600 * 10**9 / factor
```
Comment (jbrockmendel) [inline pandas/_libs/tslibs/period.pyx:889]: so the trouble is that there is an overflow going on somewhere in this expression?

Comment (WillAyd) [reply pandas/_libs/tslibs/period.pyx:889]: Yea I am fairly certain that `24 * 3600 * 10**9` will overflow - these are likely interpreted by the compiler to just be of type int, but that multiplication could very well exceed the limits of an int type. Adding the ULL suffix I think would be ideal 

More details on how decimal literals are assigned types here:
https://stackoverflow.com/a/41407498/621736

Comment (nrebena) [reply pandas/_libs/tslibs/period.pyx:889]: Using uint64_t instead of int64_t will work for the given example, but will then fail for date range earlier than the epoch with an integer overflow, so we must stick with signed integer here.
About how this change work, it just change the order of the operation so that unix_date is not multiplied by `24 * 3600 * 10**9`, but by `24 * 3600 * 10**9 / factor`, which is smaller and does not result into an integer overflow (except for value in the really far futur for the use case described in the issue, after the year `2*10**15`)
So the real fix to do here is maybe just to add parenthesis in the right place, see new commit shortly.

File: doc/source/whatsnew/v1.1.3.rst (hunk 1 above)
Comment (nrebena) [reply doc/source/whatsnew/v1.1.3.rst]: Sure thing.

File: pandas/_libs/tslibs/period.pyx (hunk 2 above)
Comment (nrebena) [reply pandas/_libs/tslibs/period.pyx]: Do you mean declaring this way: `cdef const nanos_in_day = 24 * 3600 * 10**9` ?
This does not compile.

**Comment by nrebena** [comment]
@jreback @WillAyd green and comment addressed, even if I am not sure about the const comment.

I also explicited in response to comment how the fix work, and why we should not be using unsigned long long here.

**Review by jreback — APPROVED** [review]

File: pandas/_libs/tslibs/period.pyx (hunk 3 above)
Comment (jreback) [reply pandas/_libs/tslibs/period.pyx:889]:
```
In [64]: 24 * 3600 * 10**9                                                                                                                                
Out[64]: 86400000000000

In [65]: np.iinfo(np.int64).max                                                                                                                           
Out[65]: 9223372036854775807

```

this is a fairly standard number, i agree if multiplied by a large number this *could* overflow, but ok here.

**Comment by jreback** [comment]
@WillAyd anything further?

**Review by WillAyd — APPROVED** [review]

**Comment by WillAyd** [comment]
Great thanks @nrebena - nice PR!

**Comment by simonjayhawkins** [comment]
@meeseeksdev backport 1.1.x
